src/tracing.ts

```diff
@@ -0,0 +1,166 @@
+}
+
+// PluginLoaderConfig extends TraceAgentConfig
+export type NormalizedConfig =
+    (TraceWriterConfig&PluginLoaderConfig&TopLevelConfig)|{enabled: false};
+
+/**
+ * A class that represents automatic tracing.
+ */
+export class Tracing implements Component {
+  /** A logger. */
+  private readonly logger: common.Logger;
+  /** The configuration object for this instance. */
+  private readonly config: Forceable<NormalizedConfig>;
+
+  /**
+   * Constructs a new Tracing instance.
+   * @param config The configuration for this instance.
+   * @param traceAgent An object representing the custom tracing API.
+   */
+  constructor(
+      config: NormalizedConfig, private readonly traceAgent: TraceAgent) {
+    this.config = config;
+    this.traceAgent = traceAgent;
+    let logLevel = config.enabled ? config.logLevel : 0;
+    // Clamp the logger level.
+    if (logLevel < 0) {
+      logLevel = 0;
+    } else if (logLevel >= common.logger.LEVELS.length) {
+      logLevel = common.logger.LEVELS.length - 1;
+    }
+    this.logger = common.logger({
+      level: common.logger.LEVELS[logLevel],
+      tag: '@google-cloud/trace-agent'
+    });
+  }
+
+
+
+  /**
+   * Logs an error message detailing the list of modules that were loaded before
+   * the Trace Agent. Loading these modules before the Trace Agent may prevent
+   * us from monkeypatching those modules for automatic tracing.
+   * @param filesLoadedBeforeTrace The list of files that were loaded using
+   * require() before the Stackdriver Trace Agent was required.
+   */
+  logModulesLoadedBeforeTrace(filesLoadedBeforeTrace: string[]) {
+    const modulesLoadedBeforeTrace: string[] = [];
+    const traceModuleName = path.join('@google-cloud', 'trace-agent');
+    for (let i = 0; i < filesLoadedBeforeTrace.length; i++) {
+      const moduleName = packageNameFromPath(filesLoadedBeforeTrace[i]);
+      if (moduleName && moduleName !== traceModuleName &&
+          modulesLoadedBeforeTrace.indexOf(moduleName) === -1) {
+        modulesLoadedBeforeTrace.push(moduleName);
+      }
+    }
+    if (modulesLoadedBeforeTrace.length > 0) {
+      this.logger.error(
+          'TraceAgent#start: Tracing might not work as the following modules',
+          'were loaded before the trace agent was initialized:',
+          `[${modulesLoadedBeforeTrace.sort().join(', ')}]`);
+    }
+  }
+
+  /**
+   * Enables automatic tracing support and the custom span API.
+   */
+  enable(): void {
+    if (!this.config.enabled) {
+      return;
+    }
+
+    try {
+      // Initialize context propagation mechanism.
+      const clsConfig: Forceable<TraceCLSConfig> = {
+        mechanism: this.config.clsMechanism as TraceCLSMechanism,
+        [FORCE_NEW]: this.config[FORCE_NEW]
+      };
+      cls.create(clsConfig, this.logger).enable();
+
+      traceWriter.create(this.config, this.logger).initialize((err) => {
+        if (err) {
+          this.disable();
+        }
+      });
+    } catch (e) {
```

Comment:
I still think the error handling flow here needs to be explained. You're interested in catching and logging the synchronous errors throws by `traceWriter.create` but not the async ones given by `initialize`? Why the difference?

This suggests to me that `initialize` is not expected to throw. If so, why is it in the try block?

Aside: this code could be simplified in a follow-on if `initialize` was converted to an async function.

Reply:
`initialize` is indeed not expected to throw. We can move it out of the `try` block.

What is the downside of keeping all of the logic together in the `try` block? The implementation of this function (`enable`) shouldn't depend on whether it's possible for a function to throw -- because those functions might have new conditions to throw later. I think trying to minimize the `try` block as much as possible leads to unneeded complexity in reading the code.